D2M Pass 4: Tensor Layout

jdesousa-TT opened the pull request:
### Ticket
#1908 

### Problem description
This is a complex problem. The tensor layout pass will evolve over time to make better and more efficient decisions.

### What's changed
This initial iteration of the pass will do four things:
1. Rewrite the function return to add a ~~`reinterpret_layout`~~ `to_layout` from whatever layout the final op produces, to the expected output of the original function. Insert an empty tensor op to hold this result, and rewrite the function return op.
2. Rewrite the outputs of each ttir generic op to use the optimal tensor layout. The initial heuristic is as follows:
    - Maximize the grid shape that the tensor will use
    - ~~Minimize the padding needed to achieve this size~~
    - Never allow for entire cores to be working on padding alone 

   This section will also rewrite the worker grid attribute of the generic to match the output grid. 
3. Rewrite the function operands to their optimal layouts. This does not change the function operands themselves, it only changes the uses of these operands to be references to ~~`reinterpret_layout`~~ `to_layout`s and allocate the empty tensor ops required to store this transformation.
4. Rewrite generic block memrefs to match the new tensor layouts. 

### Checklist
- [x] New/Existing tests provide coverage for changes

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,243 @@
+        signalPassFailure();
+        return;
+      }
+    }
```
Do these passes need strict ordering? Why can't they be grouped together?

jdesousa-TT replied:
At some point there was more dependence on order, now just the memref rewriter pass needs to go last, so I've grouped the first three.

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,243 @@
+    auto dpsOp = mlir::cast<DestinationStyleOpInterface>(op);
+    assert(dpsOp.getNumDpsInits() == 1 &&
+           "Only one result tensor is supported for now");
+    dpsOp.getDpsInits()[0].setType(optimal_layout);
```
I don't think it's legal to use `set*` operations inside of a pattern rewriter.  I think all graph modifications must happen inside of `PatternRewriter` so these need to be wrapped in `rewriter.modifyOpInPlace`

jdesousa-TT replied:
Wrapped all uses of the `set*` operations inside of `modifyOpInPlace` callbacks.

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,243 @@
+  auto result_encoding =
+      mlir::dyn_cast_or_null<MetalLayoutAttr>(result_type.getEncoding());
+  assert(result_encoding && "Tensor type must have a MetalLayoutAttr encoding");
+  auto optimal_output_grid = getOptimalGrid(
```
Probably we need to assert that this layout's grid shape is all 1's since we're assuming here that the shard shape is the fully expanded physical/tiled shape.

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,243 @@
+        mlir::cast<MetalLayoutAttr>(optimal_layout.getEncoding()).getGrid());
+
+    return failure(); // need some better way to exit cond. the rewriter than
+                      // always returning false!
```
Return success if you want to commit your changes and then recursively run the pattern.  Return failure if you don't want to commit.

Typically you need to test at the top to see if your change has already been applied, esp in this case where you're not changing the op type so this pattern will be matched for every recursive step.

jdesousa-TT replied:
Updated the success/failure logic on most of the passes. Please take another look.

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,243 @@
+    }
+    if (grid_shape.size() == i + 1) {
+      continue;
+    }
```
I think let's not worry about pad for now, we don't have a way of implementing that kind of gather pattern yet and we have more fundamental things to get first.  Let's just pick the first largest grid that divides memref dim[I]

jdesousa-TT replied:
I'm not sure if this interferes with our current padding support at all. In the case that the grid evenly divides the tensor, it will always pick that largest divisor. The padding here just safeguards us for odd shapes, and will essentially "round up" the tensor size to fit into tile aligned memrefs that are divisible by the grid. If this isn't what we want at this stage can you tell me a bit more about how we should handle odd tensor shapes for the interim?

nsmithtt replied:
> I'm not sure if this interferes with our current padding support at all

I think this is what I'm getting at, we don't really have padding support anywhere else, so it's unclear if we're ready to lock in on this heuristic.

Nit we should use camelCase.

It might be written more succinctly:
```c++
for (size_t i = 0; i < memrefShape.size(); i++) {
    int64_t dim = memrefShape[i];
    int64_t gridDim = deviceGridShape[i];
    int64_t shardDim = llvm::divideCeil(dim, gridDim);
    gridShape.push_back(llvm::divideCeil(dim, shardDim));
}
```

jdesousa-TT replied:
Just to be clear, should I just assert out for shapes that are not tile aligned already, and then we can just divide the grid? Even if the tensor is tile aligned, take something like `19x1` tiles, should the grid shape just remain `1x1` because we can't pad out to a divisor of the grid?

nsmithtt replied:
Yeah we'll have to align up to tile for sure.  Yeah I was thinking just picking even divisors for now, which I understand is unfortunate for primes > 8.  Maybe it doesn't matter either way, I guess it's fine if we do it this way for now, I think I didn't fully understand how the pad was working here at first, but I think we should be flexible to change it if we need to in the future.

nsmithtt reviewed (commented):
It's off to a good start!  Some comments inline

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,234 @@
+    assert(op->getResults().size() == 1 &&
+           "Only one result tensor is supported for now");
+    auto optimal_layout = getLocalLayout(op->getResult(0), rewriter, device);
+    if (genericOp.getGridAttr() != GridAttr::get(rewriter.getContext()) ||
```
Is this condition ever true if the latter condition isn't?

jdesousa-TT replied:
It could be on the second iteration of the rewrite.

Take this example 1x1 grid, 8x12 tiles.
Optimal layout will be 8x6 grid, 1x2 tiles.

On the second iteration:
New optimal layout will be based on 1x2 tiles since it's been rewritten, so optimal layout is 1x2 grid, 1x1 tiles per core.
But since it's been rewritten already and grid != 1x1, we fail here.

nsmithtt replied:
> New optimal layout will be based on 1x2 tiles since it's been rewritten, so optimal layout is 1x2 grid, 1x1 tiles per core.

Once we canonicalize the shape by multiplying out the grid shape as suggested above, we can change this to be a single check `genericOp.getResult(0).getType() == optimalLayout`.  Perhaps we should rename `optimalLayout` to `newTensorType` since it's a tensor type and not a layout.

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,234 @@
+      mlir::dyn_cast_or_null<MetalLayoutAttr>(result_type.getEncoding());
+  assert(result_encoding && "Tensor type must have a MetalLayoutAttr encoding");
+  auto optimal_output_grid = getOptimalGrid(
+      tensor.getContext(), result_encoding.getMemref().getShape(),
```
I think `result_encoding.getMemref().getShape()` might be clearer as `llvm::divideCeil(resultEncoding.getPhysicalShape(...), 32)`.

jdesousa-TT replied:
I'll add an assertion above this to ensure that the memref element type is tiled. Then it should be clear the memref shape represents tiles.

jdesousa-TT commented:
Hello @nsmithtt @xanderchin @vroubtsovTT,

I'm asking for one more look on this PR. As discussed, I have removed the ability to pad for the sake of maximizing the grid for now, until we have a more concrete way of handling this.

Fixed some other feedback from above.

Added a test case for a large prime tensor which must go on 1x1 grid since we aren't padding at this stage.

nsmithtt commented on include/ttmlir/Dialect/TTIR/Transforms/Passes.td:
```diff
@@ -38,6 +38,14 @@ def TTIRAttachMetalLayout: Pass<"ttir-attach-metal-layout", "::mlir::ModuleOp">
+def TTIRTensorLayout: Pass<"ttir-tensor-layout", "::mlir::ModuleOp"> {
+  let summary = "";
+  let description = [{
+    Analyze the graph and select optimal layouts, insert reinterpret_layout where needed.
```
`reinterpret_` -> `view_`

jdesousa-TT replied:
It's actually just `to_layout` now :)

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,226 @@
+  return GridAttr::get(ctx, gridShape);
+}
+
+RankedTensorType getLocalLayout(Value tensor, PatternRewriter &rewriter,
```
Function should be static and typically the `PatternRewriter` should be the first argument.  Let's also name this, `calculateOptimalLayoutForTensorType`.

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,226 @@
+#define GEN_PASS_DEF_TTIRTENSORLAYOUT
+#include "ttmlir/Dialect/TTIR/Transforms/Passes.h.inc"
+
+GridAttr getOptimalGrid(MLIRContext *ctx, ArrayRef<int64_t> memrefShape,
```
Function should be `static`

jdesousa-TT replied:
Encapsulated all of these in an anon namespace. Seems to be preferred over `static`.

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,226 @@
+RankedTensorType getLocalLayout(Value tensor, PatternRewriter &rewriter,
+                                DeviceAttr &device) {
+  RankedTensorType resultType =
+      mlir::dyn_cast<RankedTensorType>(tensor.getType());
```
If we're expecting to immediately dereference, we should use `mlir::cast` instead of `dyn_cast`.

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,226 @@
+  auto resultEncoding =
+      mlir::dyn_cast_or_null<MetalLayoutAttr>(resultType.getEncoding());
+  assert(resultEncoding && "Tensor type must have a MetalLayoutAttr encoding");
+  assert(mlir::isa<TileType>(resultEncoding.getElementType()) && "Inputs to the tensor layout pass must be tiled.");
```
I'm not sure that this is true.  We will need support for row-major data in order to upload data initially and read back from device.

jdesousa-TT replied:
Removed, this isn't really necessary anymore since we are just splitting work evenly and not padding. Therefore the split logic should work either way with row major or tiled data.

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,226 @@
+      }
+    }
+  }
+  return GridAttr::get(ctx, gridShape);
```
I think we should pass the rewriter here so we can write:
```c++
return rewriter.getType<GridAttr>(gridShape);
```

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,226 @@
+                           resultEncoding.getMemorySpace(),
+                           optimalOutputGrid)
+          .withElementType(tensor.getContext(),
+                           resultEncoding.getMemref().getElementType());
```
I think this might be better written:
```c++
auto newResultEncoding = resultEncoding.withGrid(resultType, optimalOutputGrid);
```

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,226 @@
+                           resultEncoding.getMemref().getElementType());
+
+  auto newTensorType =
+      RankedTensorType::get(resultType.getShape(),
```
Can probably just return directly.

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,226 @@
+  assert(resultEncoding && "Tensor type must have a MetalLayoutAttr encoding");
+  assert(mlir::isa<TileType>(resultEncoding.getElementType()) && "Inputs to the tensor layout pass must be tiled.");
+  auto optimalOutputGrid = getOptimalGrid(
+      tensor.getContext(), resultEncoding.getMemref().getShape(),
```
We need to multiply out the grid shape to canonicalize it i.e. this should be `resultEncoding.getMemref().getShape() * resultEncoding.getGrid().getShape()`.

jdesousa-TT replied:
Changed this to canonicalize like above.

Q: Is there a case where grid rank != memref rank? If so, do we just pad the front of the smaller rank with 1s to multiply?

nsmithtt replied:
> Is there a case where grid rank != memref rank? 

They must be equal in rank, it should be safe to assert this.

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,226 @@
+    func::FuncOp funcOp = mlir::cast<func::FuncOp>(op);
+
+    rewriter.setInsertionPointToStart(&op->getRegion(0).front());
+    ;
```
Stray `;`

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,226 @@
+                  .getShape() == encoding.getGrid().getShape() ||
+          (std::distance(operand.getUses().begin(), operand.getUses().end()) ==
+               1 &&
+           mlir::isa<ttir::ToLayoutOp>(operand.getUses().begin().getUser()))) {
```
Shouldn't this just be `tensor.getType() == optimalLayout`?  I don't think it should be checking if there's a to_layout user because it's not guaranteed to be going into a to_layout, nor is it guaranteed that it's not a spurious to_layout that's doing format conversion or something else.

jdesousa-TT replied:
It can't be `tensor.getType() == optimalLayout` because I'm specifically not changing the function signature, but instead inserting to_layouts where the function inputs/outputs are not optimal, using the optimal layout for computation, then relayout the output so the function return is satisfied. 

Do you think it is okay to just change the function signature directly? Not sure how this will affect a model for example.

Example:

```
#layout1 = #tt.metal_layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<8x12x!tt.tile<32x32, f32>, #l1_>>
#layout2 = #tt.metal_layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<8x1x!tt.tile<32x32, f32>, #l1_>>

func.func @reduce_large_grid(%arg0: tensor<256x384xf32, #layout1>, %arg1: tensor<256x384xf32, #layout1>) -> tensor<256x32xf32, #layout2> {
  %0 = tensor.empty() : tensor<256x32xf32, #layout2>
  %1 = "ttir.generic"(%arg0, %arg1, %0) <{
        grid = #tt.grid<1x1>,
        indexing_maps = [#map1, #map1, #map2],
        iterator_types = [#parallel, #reduction],
        operandSegmentSizes = array<i32: 2, 0, 1>
        }> ({
        ^bb0(%arg2: memref<8x12x!tt.tile<32x32, f32>, #l1_>,
            %arg3: memref<8x12x!tt.tile<32x32, f32>, #l1_>,
            %arg4: memref<8x1x!tt.tile<32x32, f32>, #l1_>):
            linalg.generic {
                indexing_maps = [#map1, #sc_map, #map2],
                iterator_types = ["parallel", "parallel"]}
                ins(%arg2, %arg3: memref<8x12x!tt.tile<32x32, f32>, #l1_>, memref<8x12x!tt.tile<32x32, f32>, #l1_>)
                outs(%arg4: memref<8x1x!tt.tile<32x32, f32>, #l1_>) {
                ^bb0(%a: !tt.tile<32x32, f32>, %b: !tt.tile<32x32, f32>, %c: !tt.tile<32x32, f32>):
                    %2 = "ttir.tile_reduce_max" (%a, %b) {reduce_dim = #ttir<reduce_dim R>} : (!tt.tile<32x32, f32>, !tt.tile<32x32, f32>) -> !tt.tile<32x32, f32>
                    linalg.yield %2: !tt.tile<32x32, f32>
            }
        "ttir.yield"() : () -> ()
        }) : (tensor<256x384xf32, #layout1>, tensor<256x384xf32, #layout1>, tensor<256x32xf32, #layout2>) -> tensor<256x32xf32, #layout2>
  return %1 : tensor<256x32xf32, #layout2>
}
```

**currently becomes:**

```
#layout = #tt.metal_layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<8x12x!tt.tile<32x32, f32>, #l1_>>
#layout1 = #tt.metal_layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<8x1x!tt.tile<32x32, f32>, #l1_>>
#layout2 = #tt.metal_layout<(d0, d1) -> (d0, d1), undef, <8x6>, memref<1x2x!tt.tile<32x32, f32>, #l1_>>
#layout3 = #tt.metal_layout<(d0, d1) -> (d0, d1), undef, <8x1>, memref<1x1x!tt.tile<32x32, f32>, #l1_>>

  func.func @reduce_large_grid(%arg0: tensor<256x384xf32, #layout>, %arg1: tensor<256x384xf32, #layout>) -> tensor<256x32xf32, #layout1> {
    %0 = tensor.empty() : tensor<256x384xf32, #layout2>
    %1 = "ttir.to_layout"(%arg0, %0) : (tensor<256x384xf32, #layout>, tensor<256x384xf32, #layout2>) -> tensor<256x384xf32, #layout2>
    %2 = tensor.empty() : tensor<256x384xf32, #layout2>
    %3 = "ttir.to_layout"(%arg1, %2) : (tensor<256x384xf32, #layout>, tensor<256x384xf32, #layout2>) -> tensor<256x384xf32, #layout2>
    %4 = tensor.empty() : tensor<256x32xf32, #layout3>
    %5 = "ttir.generic"(%1, %3, %4) <{grid = #tt.grid<8x1>, indexing_maps = [#map, #map, #map2], iterator_types = [#parallel, #reduction], operandSegmentSizes = array<i32: 2, 0, 1>}> ({
    ^bb0(%arg2: memref<1x2x!tt.tile<32x32, f32>, #l1_>, %arg3: memref<1x2x!tt.tile<32x32, f32>, #l1_>, %arg4: memref<1x1x!tt.tile<32x32, f32>, #l1_>):
      linalg.generic {indexing_maps = [#map, #map1, #map2], iterator_types = ["parallel", "parallel"]} ins(%arg2, %arg3 : memref<1x2x!tt.tile<32x32, f32>, #l1_>, memref<1x2x!tt.tile<32x32, f32>, #l1_>) outs(%arg4 : memref<1x1x!tt.tile<32x32, f32>, #l1_>) {
      ^bb0(%in: !tt.tile<32x32, f32>, %in_0: !tt.tile<32x32, f32>, %out: !tt.tile<32x32, f32>):
        %8 = "ttir.tile_reduce_max"(%in, %in_0) <{reduce_dim = #ttir<reduce_dim R>}> : (!tt.tile<32x32, f32>, !tt.tile<32x32, f32>) -> !tt.tile<32x32, f32>
        linalg.yield %8 : !tt.tile<32x32, f32>
      }
      "ttir.yield"() : () -> ()
    }) : (tensor<256x384xf32, #layout2>, tensor<256x384xf32, #layout2>, tensor<256x32xf32, #layout3>) -> tensor<256x32xf32, #layout3>
    %6 = tensor.empty() : tensor<256x32xf32, #layout1>
    %7 = "ttir.to_layout"(%5, %6) : (tensor<256x32xf32, #layout3>, tensor<256x32xf32, #layout1>) -> tensor<256x32xf32, #layout1>
    return %7 : tensor<256x32xf32, #layout1>
  }
```

**If we change the function args directly:**

```
#layout2 = #tt.metal_layout<(d0, d1) -> (d0, d1), undef, <8x6>, memref<1x2x!tt.tile<32x32, f32>, #l1_>>
#layout3 = #tt.metal_layout<(d0, d1) -> (d0, d1), undef, <8x1>, memref<1x1x!tt.tile<32x32, f32>, #l1_>>

func.func @reduce_large_grid(%arg0: tensor<256x384xf32, #layout2>, %arg1: tensor<256x384xf32, #layout2>) -> tensor<256x32xf32, #layout3> {
  %4 = tensor.empty() : tensor<256x32xf32, #layout3>
  %5 = "ttir.generic"(%arg0, %arg1, %4) <{grid = #tt.grid<8x1>, indexing_maps = [#map, #map, #map2], iterator_types = [#parallel, #reduction], operandSegmentSizes = array<i32: 2, 0, 1>}> ({
  ^bb0(%arg2: memref<1x2x!tt.tile<32x32, f32>, #l1_>, %arg3: memref<1x2x!tt.tile<32x32, f32>, #l1_>, %arg4: memref<1x1x!tt.tile<32x32, f32>, #l1_>):
    linalg.generic {indexing_maps = [#map, #map1, #map2], iterator_types = ["parallel", "parallel"]} ins(%arg2, %arg3 : memref<1x2x!tt.tile<32x32, f32>, #l1_>, memref<1x2x!tt.tile<32x32, f32>, #l1_>) outs(%arg4 : memref<1x1x!tt.tile<32x32, f32>, #l1_>) {
    ^bb0(%in: !tt.tile<32x32, f32>, %in_0: !tt.tile<32x32, f32>, %out: !tt.tile<32x32, f32>):
      %8 = "ttir.tile_reduce_max"(%in, %in_0) <{reduce_dim = #ttir<reduce_dim R>}> : (!tt.tile<32x32, f32>, !tt.tile<32x32, f32>) -> !tt.tile<32x32, f32>
      linalg.yield %8 : !tt.tile<32x32, f32>
    }
    "ttir.yield"() : () -> ()
  }) : (tensor<256x384xf32, #layout2>, tensor<256x384xf32, #layout2>, tensor<256x32xf32, #layout3>) -> tensor<256x32xf32, #layout3>
  return %4 : tensor<256x32xf32, #layout3>
}
```

Is this safe to change? @nsmithtt 

Otherwise, I need some way to determine that this input is either already optimal, or that to_layouts have been inserted in place of all original uses for consumption by compute.

jdesousa-TT replied:
This has been changed in the latest commit. Resolved.

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,226 @@
+  return newTensorType;
+}
+
+class TTIRGenericTensorLayoutRewriter : public RewritePattern {
```
All of these might be better as an `OpRewritePattern<...>`

azecevicTT commented on test/ttmlir/Dialect/TTIR/tensor_layout/tensor_layout.mlir:
```diff
@@ -0,0 +1,80 @@
+        operandSegmentSizes = array<i32: 2, 0, 1>
+        }> ({
+        // CHECK: ^bb0(%arg2: memref<1x2x!tt.tile<32x32, f32>, #l1_>,
+        // CHECK: %arg3: memref<1x2x!tt.tile<32x32, f32>, #l1_>,
```
`CHECK-SAME` should be preferred here, because it guarantees that `%arg3` will be on the same line as `%arg2`, `CHECK` only guarantees that it will appear somewhere in the IR. Also you don't need to match the text contagiously, though you might prefer `,` here as some form of guarantee that there is something after it.

jdesousa-TT replied:
Changed to use CHECK-SAME on subsequent lines.

azecevicTT reviewed (approved):
Approve for the non-TTMetal changes.

vroubtsovTT commented on test/ttmlir/Dialect/TTIR/tensor_layout/tensor_layout.mlir:
```diff
@@ -0,0 +1,80 @@
+        grid = #tt.grid<1x1>,
+        indexing_maps = [#map1, #map1, #map2],
+        iterator_types = [#parallel, #reduction],
+        operandSegmentSizes = array<i32: 2, 0, 1>
```
somehow I expect the correct sizing here to be `operandSegmentSizes = array<i32: 2, 1>`: 2 inputs and 1 outout.  `ttir.generic` no longer has CB operands.

vroubtsovTT commented on test/ttmlir/Dialect/TTIR/tensor_layout/tensor_layout.mlir:
```diff
@@ -0,0 +1,80 @@
+        grid = #tt.grid<1x1>,
+        indexing_maps = [#map1, #map1, #map2],
+        iterator_types = [#parallel, #reduction],
+        operandSegmentSizes = array<i32: 2, 0, 1>
```
same comment about `operandSegmentSizes = array<i32: 2, 1>`

jdesousa-TT replied:
Fixed

vroubtsovTT commented on lib/Conversion/TTIRToTTMetal/CMakeLists.txt:
```diff
@@ -2,6 +2,7 @@ add_mlir_conversion_library(TTMLIRTTIRToTTMetal
   TTIRToTTMetal.cpp
   TTIRToTTMetalPass.cpp
   AttachMetalLayout.cpp
+  TensorLayout.cpp
```
I would prefer if we stuck to the style of naming passes as the action/verb of what they do:

- we have AttachMetalLayout
- I am working on GeneralizeNamedOps
- so perhaps `SelectTensorLayout` here then?

or we could also go the other way and use nouns (TensorLayout, MetalLayout, Generalization, etc) -- but perhaps we should be consistent

jdesousa-TT replied:
I like the verb, I think it's more descriptive of what is actually happening. How does `OptimizeTensorLayout` sound?  Just because technically there's already a tensor layout before this pass.

vroubtsovTT replied:
'OptimizeTensorLayout' sounds good to me

vroubtsovTT commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,226 @@
+    }
+
+    rewriter.modifyOpInPlace(genericOp, [&]() {
+      genericOp->getResult(0).setType(optimalLayout);
```
I am looking at ttir.generic signature and see that its result type is a variadic:
```
def TTIR_GenericOp : TTIR_BufferizableOp<"generic", [AttrSizedOperandSegments, NoTerminator]> {
...
    let results = (outs Variadic<AnyRankedTensor>:$results);
```
so perhaps hardcoding to only `getResult(0)` would be wrong here? At the very least, the assert about size of DPS input being 1 that follows perhaps should be moved to the start of the lamba.

jdesousa-TT replied:
I think an assertion is appropriate for now. I don't know if we have / plan to support any ops with >1 outputs?

vroubtsovTT reviewed (changes requested):
some style changes for sure, anonymous ns has to be done or we can have painful ODR bugs again

jdesousa-TT commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,226 @@
+                       ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
```
Funnily enough, I've been doing this in the D2M backend and everything new. I wrote these decls a long time ago, before all of our discussion about types of rewriters, etc. Updated all of these, much more succinct.

jdesousa-TT commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,226 @@
+
+class TTIRMemrefLayoutRewriter : public RewritePattern {
+public:
+  TTIRMemrefLayoutRewriter(MLIRContext *ctx)
```
Conventionally in the project we seem to have always used classes, but I suppose you're right - there's no need for these rewrite patterns to be classes. Should we follow what we've already been doing for consistency, or use structs here?

vroubtsovTT replied:
I have always seen 'struct' as an alias to "all public" 'struct', so I don't have any mental hangups with either. Stan Lippman has a funny story about this in his "Inside the C++ Object Model":

>Surprisingly, one of the most-asked questions of
C programmers moving to C++ (once the questions about performance have been put aside) is when, if ever,
should one use a struct declaration rather than a class declaration when writing a program using C++?
...
>At that time, this was one of those small philosophical points that one necessarily
does not point out. However, it is one from which one gains some small (admittedly quite small) satisfaction
when, well, when it is pointed out, usually as a question: "Hey, did you know? Struct. The keyword. It's not
used anywhere.…" But as a colleague of mine at Bell Laboratories indirectly pointed out to me, even the
smallest philosophical point can have a human cost. For example, a C programmer in his group, anxious to
learn C++, was quite distressed to find struct absent from my book. Apparently its inclusion would have
provided a transitional lifeline to make the programmer's rocky ascent less bruising. So much, then, for
philosophy.

On a more series note, I've checked llvm source tree and there are 660 matches for `struct.*final` vs 1911 for `class.*final`, so both are used a lot. I would be fine with either but would prefer `struct` for a small class that has nothing but one public method and `class` for something more serious.

jdesousa-TT commented on lib/Conversion/TTIRToTTMetal/TensorLayout.cpp:
```diff
@@ -0,0 +1,226 @@
+                  .getGrid()
+                  .getShape() == encoding.getGrid().getShape() ||
+          (std::distance(operand.getUses().begin(), operand.getUses().end()) ==
+               1 &&
```
getUses() returns a `use_range` which doesn't have a size() member. It is an iterator range, which has start/end iterators. I think this if statement might change anyways, so may not be a problem.

vroubtsovTT replied:
Ah, I stand corrected. Note that some mlir objects provide `hasOneUse()`, not sure if that would apply to `operand`.

jdesousa-TT replied:
Hehe, this does work 🧠 , thanks!

jdesousa-TT commented on lib/Conversion/TTIRToTTMetal/OptimizeTensorLayout.cpp:
```diff
@@ -0,0 +1,237 @@
+          calculateOptimalLayoutForTensorType(rewriter, operand, device);
+      auto tensor = mlir::cast<RankedTensorType>(operand.getType());
+      auto encoding = mlir::cast<MetalLayoutAttr>(tensor.getEncoding());
+      if (mlir::cast<MetalLayoutAttr>(optimalLayout.getEncoding())
```
@nsmithtt Are there any passes before this one that would insert `to_layout`s? 
Is there a path from a front-end that would insert `to_layout`s? I'm trying to think of a scenario where this would break. I don't think it's bulletproof but I can't think of a better solution.

nsmithtt replied:
Yes, to_layout will be inserted for data format conversion or to implement some TMs.

nsmithtt replied:
Probably just simply matching the input grid shapes is OK for now, but this runs into trouble with matmul where the inner dim might not divide evenly.  Technically we need to look at the generic affine map which will tell us which iterators participate in the grid and which ones are free variables.

e.g. consider matmul affine maps:
```
(m, n, k) -> (m, k)
(m, n, k) -> (k, n)
(m, n, k) -> (m, n)
```

We must 1-1 map the output tensor's grid to the generic grid for this exact reason, because the affine maps tell us how the output grid relates to the input grids.  In this example, if we cut the output[0] dim aka `m`, then the input must also have the same divisor, by extension the same thing for `n`.  Now, `k` is also interesting, it tells us how the input's inner grid dims must relate, but `k` is decoupled from the output grid so there is more freedom in this choice.

jdesousa-TT replied:
I updated the GenericTensorLayoutRewriter to rewrite its operands along with its results. 
Removed the FuncOperands pass completely. The output is the same. This solves the problem of determining if func operands have already been lowered into to_layouts.

nsmithtt reviewed (approved):
Looks great! Awesome 🙌
